enterprise/internal/campaigns/syncer/syncer.go

```diff
@@ -160,50 +174,37 @@ func (s *SyncRegistry) handlePriorityItems() {
 	}
 }
 
-// EnqueueChangesetSyncs will enqueue the changesets with the supplied ids for high priority syncing.
-// An error indicates that no changesets have been enqueued.
-func (s *SyncRegistry) EnqueueChangesetSyncs(ctx context.Context, ids []int64) error {
-	// The channel below is buffered so we'll usually send without blocking.
-	// It is important not to block here as this method is called from the UI
-	select {
-	case s.priorityNotify <- ids:
-	default:
-		return errors.New("high priority sync capacity reached")
-	}
-	return nil
-}
-
-// HandleExternalServiceSync handles changes to external services.
-func (s *SyncRegistry) HandleExternalServiceSync(es api.ExternalService) {
-	normalised, err := externalServiceSyncerKey(es.Kind, es.Config)
+// syncCodeHosts fetches the list of currently active code hosts on the Sourcegraph instance.
+// The running syncers will then be matched against those and missing ones are spawned and
+// excess ones are stopped.
+func (s *SyncRegistry) syncCodeHosts(ctx context.Context) error {
+	codeHosts, err := s.syncStore.ListCodeHosts(ctx, store.ListCodeHostsOpts{})
 	if err != nil {
-		log15.Error(err.Error())
-		return
+		return err
 	}
 
-	s.mu.Lock()
-	syncer, exists := s.syncers[normalised]
-	s.mu.Unlock()
+	codeHostsByExternalServiceID := make(map[string]*campaigns.CodeHost)
 
-	if es.DeletedAt.IsZero() && !exists {
-		res := (types.ExternalService)(es)
-		s.Add(&res)
-	}
-
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	if !es.DeletedAt.IsZero() && exists {
-		delete(s.syncers, normalised)
-		syncer.cancel()
+	// Add and start syncers
+	for _, host := range codeHosts {
+		codeHostsByExternalServiceID[host.ExternalServiceID] = host
+		s.Add(host)
 	}
-}
 
-func externalServiceSyncerKey(kind, config string) (string, error) {
-	baseURL, err := extsvc.ExtractBaseURL(kind, config)
-	if err != nil {
-		return "", errors.Wrap(err, "getting normalized URL from service")
+	// Clean up old syncers.
+	for syncerKey := range s.syncers {
+		// If there is no code host for the syncer anymore, we want to stop it.
+		if _, ok := codeHostsByExternalServiceID[syncerKey]; !ok {
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			syncer, exists := s.syncers[syncerKey]
+			if exists {
+				delete(s.syncers, syncerKey)
+				syncer.cancel()
+			}
```

Comment:
The `defer` causes this to deadlock if the loop runs twice. You want to call `s.mu.Unlock()` at the end of this block, not at the end of the function.

```suggestion
			s.mu.Lock()

			syncer, exists := s.syncers[syncerKey]
			if exists {
				delete(s.syncers, syncerKey)
				syncer.cancel()
			}
			
			s.mu.Unlock()
```

Reply:
ooh nice, thanks!